containerd: include present content size in layer disk usage calculation

jsternberg opened the pull request:
The present content size is included in the image size usage and should
be included in the total size that the layer takes up on disk.

This prevents an issue where the reclaimable amount reported by the CLI
was a negative number.

Fixes #45395.

```markdown changelog
containerd image store: Fix `docker system df` reporting a negative reclaimable space amount
```

thaJeztah commented:
Looks like some failure that is related;

```
=== Failed
=== FAIL: amd64.integration.system TestDiskUsage/after_container.Run (0.23s)
    disk_usage_test.go:108: assertion failed: 1351680 (du.Containers[0].SizeRootFs int64) != 2114056 (du.LayersSize int64)
    --- FAIL: TestDiskUsage/after_container.Run (0.23s)
```

vvoland commented:
> Maybe I should instead add another attribute like `TotalSize` or `ContentSize` and the CLI could use that instead?

Yeah I think that would make more sense. Although that wouldn't fix the error on older CLIs that don't know about these new fields (which I think isn't the end of the world).

If we really care about fixing this on older CLI + newer server, perhaps we could make that an API versioned change:
1. Deprecate `LayersSize`
2. Make `LayersSize` include both content size + unpacked size (contrary to what the name says).
3. On API >=1.XX, introduce  `TotalSize`, `ContentSize`, `UnpackedSize`.

----

However, I also think that actually moving the "reclaimable" calculations out of the CLI, and making the server responsible for providing that information would make the most sense.

jsternberg commented:
> However, I also think that actually moving the "reclaimable" calculations out of the CLI, and making the server responsible for providing that information would make the most sense.

I like this idea. What about the following?

1. Deprecate `LayersSize`
2. Make `LayersSize` include both content size + unpacked size and make a note about it.
3. Introduce a new `DiskUsage` attribute with a possible schema.

```
{
    "DiskUsage": {
        "Images": {
            "Total": int64,
            "Reclaimable": int64
        },
        "Containers": {
            "Total": int64,
            "Reclaimable": int64
        },
        "BuildCache": {
            "Total": int64,
            "Reclaimable": int64
        }
    }
}
```

Then we can move these calculations out of the CLI and the returned attribute is a bit more extensible if we need to add something in the future.

Another alternative is to change the returned object to be something like:

```
{
    "Images": {
        "TotalSize": int64,
        "Reclaimable": int64,
        "Items": [ ... old return value ... ]
    }
}
```

This would more dramatically change the current structure though.

jsternberg commented:
Made a few different changes to this. I had one commit where I added the `TotalSize` attribute and that's the first commit. There's also a second commit with a demo of the changed structure for this endpoint and how we would handle previous API compatibility. So if you do API 1.49 you get the expanded capabilities that compute the total size and reclaimable on the server. If you use API 1.48 or below you'll get the previous structure except that `LayersSize` now includes the image manifest content for containerd.

I haven't made any changes to the swagger document or fixed any of the unit tests since I don't want to spend time on those changes if this change is just a non-starter. If this seems like a good idea, I can make those changes too.

An example of the new output.

```
{
  "Images": {
    "TotalSize": 100734776,
    "Reclaimable": 0,
    "Items": [
      {
        "Containers": 1,
        "Created": 1737951294,
        "Id": "sha256:c3d1a34325805c22bf44a5157224bcff58dc6a8868558c7746d6a2ea64eb191c",
        "Labels": {
          "org.opencontainers.image.ref.name": "ubuntu",
          "org.opencontainers.image.version": "24.04"
        },
        "ParentId": "",
        "RepoDigests": [
          "ubuntu@sha256:72297848456d5d37d1262630108ab308d3e9ec7ed1c3286a32fe09856619a782"
        ],
        "RepoTags": [
          "ubuntu:24.04"
        ],
        "SharedSize": 0,
        "Size": 100734776
      }
    ]
  },
  "Containers": {
    "TotalSize": 0,
    "Reclaimable": 0,
    "Items": [
      {
        "Id": "181ad1bff8e2679aa61c7748a57e4c7c8baef21139ae4eeef8209f3faa71b991",
        "Names": [
          "/objective_moser"
        ],
        "Image": "ubuntu:24.04",
        "ImageID": "sha256:c3d1a34325805c22bf44a5157224bcff58dc6a8868558c7746d6a2ea64eb191c",
        "Command": "/bin/bash",
        "Created": 1743435813,
        "Ports": [],
        "SizeRootFs": 100734776,
        "Labels": {
          "org.opencontainers.image.ref.name": "ubuntu",
          "org.opencontainers.image.version": "24.04"
        },
        "State": "created",
        "Status": "Created",
        "HostConfig": {
          "NetworkMode": "bridge"
        },
        "NetworkSettings": {
          "Networks": {
            "bridge": {
              "IPAMConfig": null,
              "Links": null,
              "Aliases": null,
              "MacAddress": "",
              "DriverOpts": null,
              "GwPriority": 0,
              "NetworkID": "",
              "EndpointID": "",
              "Gateway": "",
              "IPAddress": "",
              "IPPrefixLen": 0,
              "IPv6Gateway": "",
              "GlobalIPv6Address": "",
              "GlobalIPv6PrefixLen": 0,
              "DNSNames": null
            }
          }
        },
        "Mounts": []
      }
    ]
  },
  "Volumes": {
    "TotalSize": 0,
    "Reclaimable": 0,
    "Items": []
  },
  "BuildCache": {
    "TotalSize": 0,
    "Reclaimable": 0,
    "Items": []
  }
}
```

vvoland commented on api/types/types.go:
```diff
@@ -87,12 +87,38 @@ type DiskUsageOptions struct {
 // DiskUsage contains response of Engine API:
 // GET "/system/df"
 type DiskUsage struct {
-	LayersSize  int64
-	Images      []*image.Summary
-	Containers  []*container.Summary
-	Volumes     []*volume.Volume
-	BuildCache  []*BuildCache
-	BuilderSize int64 `json:",omitempty"` // Deprecated: deprecated in API 1.38, and no longer used since API 1.40.
```
We shouldn't make backwards incompatible changes to public API structs in a minor release.

jsternberg replied:
I've reverted this struct back. Is this concern about the Go API type or the HTTP response type? I can also revert the type the HTTP API returns too if that's important.

vvoland replied:
Both really. From the Go API client perspective, vendoring the client update with the struct changed would break existing code (even if it does pin the HTTP version).

On the HTTP API side, unfortunately it's a common practice for the clients to NOT pin to the exact API version, so we might need to have one API version cycle where we deprecate the old response first.

vvoland commented on daemon/containerd/service.go:
```diff
@@ -127,21 +128,58 @@ func (i *ImageService) StorageDriver() string {
+		return 0, err
+	}
+
+	visitedImages := make(map[digest.Digest]struct{})
+	for _, img := range imgs {
+		if _, ok := visitedImages[img.Target.Digest]; ok {
+			continue
+		}
+		visitedImages[img.Target.Digest] = struct{}{}
+
+		if err = i.walkReachableImageManifests(ctx, img, func(imageManifest *ImageManifest) error {
+			size, err := imageManifest.PresentContentSize(ctx)
+			if err != nil {
+				return err
+			}
+			diskUsage += size
+			return nil
+		}); err != nil {
+			return 0, err
+		}
+	}
```
This will sum the same blobs multiple times if they are shared by different images.
We should walk each content blob directly and then deduplicate based on visited content blobs.

jsternberg replied:
Fixed.

thaJeztah commented on api/types/types.go:
```diff
@@ -87,12 +87,38 @@ type DiskUsageOptions struct {
+	BuildCache *BuildCacheDiskUsage
+}
+
+// ImageDiskUsage contains the image disk usage information.
+type ImageDiskUsage struct {
+	TotalSize   int64
```
I probably need to checkout this branch locally, but would it be possible to put the types in the relevant sub-packages (containers, images, ...)

We've been trying to scope these types to relevant types, because the top-level `types/` packages became a grab-bag of everything, also enforcing (indirect) imports in places where it was less than desirable.

We could have those packages imported in `system/` for use in `docker system du` , but also consider that we may have similar commands at some point for individual object-types (`docker container du` or similar), to go along with `docker container prune`.

jsternberg replied:
Doing this actually made things substantially easier. I've moved these types.

thaJeztah commented on daemon/disk_usage.go:
```diff
@@ -58,77 +72,96 @@ func (daemon *Daemon) imageDiskUsage(ctx context.Context) ([]*image.Summary, err
 // SystemDiskUsage returns information about the daemon data disk usage.
 // Callers must not mutate contents of the returned fields.
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (du *types.DiskUsage, err error) {
```
Try to avoid named output variables; I don't think we need them in this case. We try to avoid them and only use them in exceptional cases (usually if there's an error to be handled in a `defer`, and in that case we make sure it's not named `err`, but (e.g.) `retErr` to make it less err-prone (we've had various cases where `err` was accidentally shadowed).

jsternberg commented on api/types/types.go:
```diff
@@ -87,12 +87,38 @@ type DiskUsageOptions struct {
+}
+
+// BuildCacheDiskUsage contains the build cache disk usage information.
+type BuildCacheDiskUsage struct {
```
Created a package for this and added a type alias.

jsternberg commented:
I've kept the internals where reclaimable and total size are calculated but kept the `/system/df` endpoint identical and the endpoint now returns the full image size in `LayersSize` so the CLI should display it correctly.

I had to remove some lines from the tests because the assumptions made in them weren't valid anymore. Since `LayersSize` now includes the content size, it can't be used to determine whether the `RepoDigests` should exist or not and that size can no longer be used to determine if `SizeRootFs` is the correct number or not since the two will differ in the containerd implementation.

This should be ready to merge and we can then determine how to use the new struct as a direct return API in the future.

vvoland commented on api/types/types.go:
```diff
@@ -84,7 +83,7 @@ type DiskUsageOptions struct {
 }
 
-// DiskUsage contains response of Engine API:
+// DiskUsage contains response of Engine API deprecated in API 1.49.
 // GET "/system/df"
```
The deprecation notice needs to be put in a separate paragraph to be correctly interpreted: https://go.dev/wiki/Deprecated

```suggestion
// DiskUsage contains response of Engine API:
// GET "/system/df"
//
// Deprecated: deprecated in API 1.49, use [system.DiskUsage] instead.
```

jsternberg replied:
I removed this deprecation note for now. The `system.DiskUsage` struct still exists but since `/system/df` is still returning it it became a bit too complicated to deprecate it since the struct still needed to be used. That would have resulted in a `nolint` in a lot of places that didn't make sense.

We can likely fix the immediate issue and do the deprecation in something different. Alternatively, we can maybe just create a new endpoint for this instead that uses the new type so we aren't causing a big inconvenience with other clients.

vvoland commented on api/types/types.go:
```diff
@@ -141,31 +140,7 @@ type BuildResult struct {
-	LastUsedAt *time.Time
-	UsageCount int
-}
+type BuildCache = build.CacheRecord
```
This should also be marked as deprecated and mention using `build.CacheRecord` directly.

vvoland commented on daemon/containerd/service.go:
```diff
@@ -127,21 +128,64 @@ func (i *ImageService) StorageDriver() string {
+
+	visitedImages := make(map[digest.Digest]struct{})
+	for _, img := range imgs {
+		if _, ok := visitedImages[img.Target.Digest]; ok {
+			continue
+		}
+
+		if err = i.walkReachableImageManifests(ctx, img, func(imageManifest *ImageManifest) error {
+			dgst := imageManifest.Target().Digest
+			if _, ok := visitedImages[dgst]; ok {
+				return nil
+			}
+			visitedImages[dgst] = struct{}{}
+
+			size, err := imageManifest.PresentContentSize(ctx)
+			if err != nil {
+				return err
+			}
+			diskUsage += size
+			return nil
+		}); err != nil {
+			return 0, err
+		}
+		visitedImages[img.Target.Digest] = struct{}{}
```
This will still count the same content twice if it's shared by multiple images.

For example, let's say I have `ubuntu` image and `my-ubuntu` built off:
```dockerfile
FROM ubuntu
RUN apt-get update && apt-get -y install fish
```

`my-ubuntu` will have two layers - one that is shared with `ubuntu` (let's call it `X`) and the second one that would be the delta produced by the `RUN apt-get ...` (let's call it `Y`).

Currently the code would count the `X` layer twice, which would not correspond to the actual disk usage because the layer blob is actually stored only once.

This needs to traverse each child content of the image and then only count the usage of the unique blobs.

jsternberg replied:
Would this be doable by just using `walkPresentChildren` on each of the images directly? It seems like that's what the internals of `PresentContentSize` is doing so I'm not sure if I need to call `walkReachableImageManifests` at all.

vvoland reviewed (approved):
LGTM, edited the changelog description a bit

dmcgowan commented on daemon/disk_usage.go:
```diff
@@ -32,7 +33,21 @@ func (daemon *Daemon) containerDiskUsage(ctx context.Context) ([]*container.Summ
-		return containers, nil
+
+		isActive := func(ctr *container.Summary) bool {
+			return strings.Contains(ctr.State, "running") ||
```
Should be updated for #49965

jsternberg replied:
Rebased and updated.

dmcgowan reviewed (approved):
The update to use consts could be here or in a follow up, code wise looks good

vvoland reviewed (changes requested):
lint failure:
```
 api/server/router/system/backend.go:46:33: SA1019: types.BuildCache is deprecated: deprecated in API 1.49. Use [build.CacheRecord] instead. (staticcheck)
	DiskUsage(context.Context) ([]*types.BuildCache, error)
```

jsternberg committed:
containerd: include present content size in disk usage calculations

The present content size is included in the image size usage and should
be included in the total size that the layer takes up on disk.

This prevents an issue where the reclaimable amount reported by the CLI
was a negative number.

This also updates the `/system/df` endpoint to use a new type that
computes information that was previously computed by the CLI. Computing
these in the server should require less work from the CLI and ensure
the calculations are more accurate because the CLI doesn't have to
reconstruct the numbers.

Signed-off-by: Jonathan A. Sternberg <jonathan.sternberg@docker.com>

jsternberg commented:
Fixed the lint failure.
